diesel_infer_schema/src/codegen.rs

```diff
@@ -4,10 +4,252 @@ use quote;
 use syn;
 
 use table_data::TableData;
-use data_structures::ColumnInformation;
+use data_structures::{ColumnInformation, EnumInformation};
 use inference::{establish_connection, get_table_data, determine_column_type, get_primary_keys,
                 InferConnection};
 
+pub fn wrap_item_in_const(const_name: syn::Ident, item: quote::Tokens) -> quote::Tokens {
+    quote! {
+        #[allow(dead_code)]
+        const #const_name: () = {
+            extern crate diesel;
+            #item
+        };
+    }
+}
+
+fn to_uppercase(ty: String) -> String {
+    let mut ret = String::with_capacity(ty.len());
+    let mut next_uppercase = true;
+    for c in ty.chars() {
+        if c == '_' {
+            next_uppercase = true;
+            continue;
+        }
+        if next_uppercase {
+            ret += &c.to_uppercase().to_string();
+        } else {
+            ret.push(c);
+        }
+        next_uppercase = false;
+    }
+    ret
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum ExpandEnumMode {
+    PrettyPrint,
+    Codegen,
+}
+
+pub fn expand_enum(enum_info: EnumInformation, mode: ExpandEnumMode) -> quote::Tokens {
+    let EnumInformation{type_name, fields} = enum_info;
+    let sql_mod = syn::Ident::new(format!("sql_{}", type_name));
+    let sql_type_name = syn::Ident::new(to_uppercase(format!("{}Sql", type_name)));
+    let type_name_string = type_name.clone();
+    let type_name = syn::Ident::new(to_uppercase(type_name));
+
+    let field_mapping = fields.clone().into_iter().map(|field| {
+        let ident = syn::Ident::new(to_uppercase(field.clone()));
+        quote!(#field => #type_name::#ident)
+    }).collect::<Vec<_>>();
+    let reverse_mapping = fields.clone().into_iter().map(|field| {
+        let ident = syn::Ident::new(to_uppercase(field.clone()));
+        quote!(#type_name::#ident => #field)
+    }).collect::<Vec<_>>();
+
+    let fields = fields.into_iter().map(|field|{
+        syn::Ident::new(to_uppercase(field))
+    });
+
+        let map_names = {
+        let from_1 = {
+            let field_mapping = field_mapping.clone();
+            quote! {
+                impl<'a> From<&'a str> for #type_name {
+                    fn from(s: &'a str) -> Self {
+                        match s {
+                            #(#field_mapping,)*
+                            _ => unreachable!(),
+                        }
+                    }
+                }
+            }
+        };
+        quote!{
+            #from_1
```

Comment:
Given that the latest stable rust ups the default recursion limit to 1024, this is probably not necessary.